### DISCUSSION TITLE
Poor Performance on NAIP Imagery - Worse than just wrangling NDVI thresholding

### DISCUSSION THREAD

**Discussion by zackaryleady** [weecology/DeepForest#807] [Q&A] [answered]
Hello,

I tested this on multiple tiles of NAIP imagery by removing the NIR band on 60 cm pixel size (higher res NAIP).

Even with using a high overlap and many different patch_sizes the model performance is basically worse than just using NDVI. If I use a very small patchsize=64 then I atleast get a large number of trees where atleast a small box is found near the canopy centroid, but the entire tree or multiple connected tree crowns won't really be detected and dozens of false negatives on water pixels. IF there is an area in the tile with a large density of trees and another part of the tile has low density trees it just doesn't find the low density ones using a larger patchsize. Anything from 256-1024 really doesn't find hardly any trees even though there are hundreds-thousands of trees to find. I did check and see that you are  / 255 to normalize uint8 which is what NAIP is. Also you permute the bands to BGR internally so I didn't do that beforehand, but I did make it channels last.

I used this code:
```
def test_forest():
  model = main.deepforest()
  model.use_release()  
  with rasterio.open(r"C:\Users\zacka\Desktop\Relabel_20241007\T2\naip_w1641n1402.tif") as src:
  #with rasterio.open(r"C:\Users\zacka\Desktop\Relabel_20241007\T1\naip_w1693n1346.tif") as src:
      input_array = src.read([1, 2, 3]) # RGB

  # channels last
  input_array = input_array.transpose((1,2,0))

  #raster_path = get_data(r"C:\Users\zacka\Desktop\Relabel_20241007\T1\naip_w1693n1346.tif")
  #print(raster_path.shape)
  # Window size of 300px with an overlap of 25% among windows for this small tile.
  predicted_raster = model.predict_tile(image=input_array, iou_threshold=0.95, return_plot=True, patch_size=512 ,patch_overlap=0.90)

  # View boxes overlayed when return_plot=True, when False, boxes are returned.
  plt.imshow(predicted_raster)
  plt.show()
if __name__ == "__main__":
    test_forest()
```

### 512 patchsize / 0.9 overlap (tile is 1666x1666 pixels)
![Figure_1](https://github.com/user-attachments/assets/64e1c9d0-21dd-40ce-b151-54b11bef3ec0)
### 64 patchsize / 0.5 overlap (tile is 1666x1666 pixels)
![Figure_2](https://github.com/user-attachments/assets/5df74c44-96f5-436a-b234-4d09714f7635)

Did I miss something internally or with a different function that significantly improves performance?
I'm using mamba and version 1.3.3 of deepforest. It basically seems that NDVI, with a low threshold, then derive centroids from polygons/bounding boxes, then provide to SAM2 via samgeo, would actually outperform the performance I have seen so far. Any feedback on what I am doing wrong would be greatly appreciated.

**Comment by ethanwhite** [comment, marked as answer]
Hi @zackaryleady - welcome and thanks for the questions! We haven't done much work on data at this coarse a resolution ourselves, but [others have seen good results with NAIP data](https://doi.org/10.1016/j.jag.2024.103848). They perform two of our recommended approaches to improve on the out-of-the-box predictions, so let me start by asking:

1. Have you tried fine tuning using some hand labeled imagery for your specific sensor/region? This is the best and easiest way to get the model to behave well with data that differs from what it has been trained on: https://deepforest.readthedocs.io/en/latest/user_guide/11_training.html
2. Have you tried more systematic hyperparameter tuning? Unfortunately the above paper doesn't list the hyperparameters they found to work well on NAIP but I think a more systematic exploration of patch size (including smaller patch sizes) and iou_threshold (which you have set to be higher than we've ever used) should also help based on the paper linked above.

If you still aren't seeing decent performance at that point we can be of more help if you can provide a sample image for us to play with.

**Reply by zackaryleady** [reply]
Hello Ethan,

I really appreciate you taking the time to provide this valuable feedback. I read the paper you linked in the first hyperlink and it does seem they are able to achieve about ~70% via fine-tuning for 15 epochs on ~1,000 additional samples.

Before posting I did try a systematic search from 32-1024 on the patch size with overlap ranging from 50%-95%. I used power of 2 stepping so 32, 64, 128, 256, 512, 768, 1024. I also read the use of the iou_threshold in the NMS function from Pytorch and basically having a very high iou_threshold should allow more noise/overlap to stay  (i.e. less boxes are culled). I found regardless of this parameter, there are just a lot of trees that are not found and as the patchsize decreases which does find more trees, the boxes are smaller and smaller making the results slightly less useful. I understand that the difference between training resolution and NAIP resolution (10cm => 60 cm) is significant. I was overly optimistic. From this paper around date palm trees (which have a much more unique signature shape in RGB) the impact on ground sample size in cm is pretty significant on mIoU performance.

Figure:
![image](https://github.com/user-attachments/assets/7823c194-61fe-4f47-b59d-0c1a94c54013)

From: 
[2023-Large-ScaleDatePalmTreeSegmentationfromMultiscaleUAV-BasedandAerialImagesUsingDeepVisionTransformers.pdf](https://github.com/user-attachments/files/17382129/2023-Large-ScaleDatePalmTreeSegmentationfromMultiscaleUAV-BasedandAerialImagesUsingDeepVisionTransformers.pdf)


In terms of fine-tuning, it would seem that using some of the more recently published foundational models (Clay) and fine-tuning those with transformer based segmentation heads (Segformer) is more likely to yield better results as those foundational models are pre-trained directly on NAIP/Sent2 to begin with. If I have more time to invest in DeepForest comparisons/fine-tuning I will let you know the results. I'd be curious if your next set of research directions are for migrating to transformer based methods? There seems to be alot of interest in your work by WRI/GeoICT groups.

**Reply by ethanwhite** [reply]
Thanks for the info on your hyperparameter checks. I also [opened an issue in the repo associated with the linked paper](https://github.com/jonathanventura/urban-tree-detection/issues/3) earlier today and will be curious to see what they found to be good hyperparameters.

I agree that a model trained on NAIP/Sent2 is probably the best starting point for your use case at the moment. We are in the process of creating a much broader training set (see https://github.com/weecology/MillionTrees/) with data from different resolutions, but that is very much in the early phases. Once we've assembled that we'll be retraining the core models on it and so performance will likely be better out-of-the-box for coarser imagery in the future.

Transformers are definitely very cool technology as long as you have enough labeled data for them (or have good pre-training tasks). Adding additional (newer) architectures is definitely on the road map. Our general impression that both improvements to architecture and improvements in labeled data (including broader core datasets and more fine-tuning) are important steps in the typically label limited context of ecological applications. So, our current direction is both trying to incorporate a broader range of training data and exploring newer architectures.

**Reply by zackaryleady** [reply]
Hello @ethanwhite, any ETA when the NAIP dataset from WRI will be made available? Or do they have it somewhere already hosted? A google search didn't provide a direct link.

**Reply by jonathanventura** [reply]
Hi, as @ethanwhite mentioned we did get good results by fine-tuning DeepForest on our NAIP data.  Our data only has point annotations so we simply put a fixed size bounding box around each point and found the best box side length to be 9.6 m (16 pixels).

Since our dataset consists of 256x256 NAIP crops, we used `predict_image` instead of `predict_tile` and so we did not make use of or tune the `iou_thresh` or `patch_size` parameters.  Fine-tuning the model had the largest effect, but we did see a further boost in performance by tuning the `score_thresh` parameter -- our best value was 0.302.

Note that our dataset covers urban areas and so the trees are generally more sparsely laid out.

**Reply by zackaryleady** [reply]
Hello Jonathan,

Thank you for your reply. Would you be able to save the model weights checkpoint on your fine-tuned model? OR did I miss it on your github?

~Zack

**Reply by ethanwhite** [reply]
> any ETA when the NAIP dataset from WRI will be made available? Or do they have it somewhere already hosted? A google search didn't provide a direct link.

Spring 2025 is our ETA at the moment. A lot of the data we're compiling isn't openly available yet. Folks have been sending us data with the understanding that it will become open, but we still have to work through all of the details with individual contributors related to licensing, hosting, authorship, etc. Both @bw4sz and I are super over-committed at the moment, but it's possible that when things slow down a bit in December/January that we could try to run a fine tune for you on that data and send you the weights. Feel free to ping us in December to see if that's a possibility.

**Reply by jonathanventura** [reply]
@zackaryleady I put the DeepForest fine-tune checkpoint file [here](https://cpslo-my.sharepoint.com/:u:/g/personal/jventu09_calpoly_edu/EVPsDNupk3ZJg-QlYNn9uxoBnkxDV8z_y-OBYRBBtNS1tg?e=YI2gct) on OneDrive -- let me know if you have any issues with it.

**Reply by bw4sz** [reply]
Thanks @jonathanventura, I continue to appreciate your commitment to open data and helping out.

**Reply by markwy** [reply]
@jonathanventura , I am helping my son with a high school research project using DeepForest, and we wanted to use your checkpoint to analyze lower resolution data. However, when we tried to load it, we get errors when both DeepForest 1.5.2 and 2.0.0 are looking for "pytorch-lightning_version" (see stack snippet below). What version of the model did you fine-tune using, or is there a way (@ethanwhite ?) to use a pre-lightning checkpoint in more modern versions of deepforest? Thank you!

error in 2.0.0:
```
	File /opt/anaconda3/envs/Deepforest_1/lib/python3.11/site-packages/pytorch_lightning/utilities/migration/utils.py:143, in _pl_migrate_checkpoint(checkpoint, checkpoint_path)
    137 def _pl_migrate_checkpoint(checkpoint: _CHECKPOINT, checkpoint_path: Optional[_PATH] = None) -> _CHECKPOINT:
    138     """Applies Lightning version migrations to a checkpoint dictionary and prints infos for the user.
    139 
    140     This function is used by the Lightning Trainer when resuming from a checkpoint.
    141 
    142     """
--> 143     old_version = _get_version(checkpoint)
    144     checkpoint, migrations = migrate_checkpoint(checkpoint)
    145     new_version = _get_version(checkpoint)

File /opt/anaconda3/envs/Deepforest_1/lib/python3.11/site-packages/pytorch_lightning/utilities/migration/utils.py:164, in _get_version(checkpoint)
    162 def _get_version(checkpoint: _CHECKPOINT) -> str:
    163     """Get the version of a Lightning checkpoint."""
--> 164     return checkpoint["pytorch-lightning_version"]

KeyError: 'pytorch-lightning_version'
```

in 1.5.2:
```
File ~/PycharmProjects/PythonProject/.venv/lib/python3.10/site-packages/pytorch_lightning/utilities/migration/utils.py:164, in _get_version(checkpoint)
    162 def _get_version(checkpoint: _CHECKPOINT) -> str:
    163     """Get the version of a Lightning checkpoint."""
--> 164     return checkpoint["pytorch-lightning_version"]

KeyError: 'pytorch-lightning_version'
```
